### ISSUE TITLE
Add a CLI Interface

### ISSUE THREAD

**Issue by lucasbalena** [hoxxep/rapidhash#9]
Hi everyone,
I’m not a Rust developer, so I’m not sure how viable this idea is, but I wanted to get your thoughts.

There’s another project called [b3sum](https://crates.io/crates/b3sum) that provides a CLI tool for computing checksums using BLAKE3. For example, you can install it like this:

```shell
$ cargo install b3sum
```

And then run it on a file:
```shell
b3sum file
```

The output looks like this:
```
af46dfde5af4db763f9d2d087b4ffdbdf5ad51f4dffb53056ef42241ff144683  file
```

I use this tool in my own project, [picohash](https://github.com/lucasbalena/picohash), by calling it through a shell interface in Go:
```go
exec.Command("b3sum", path)
```

However, for my use case, BLAKE3 feels like overkill. I don’t need cryptographic security—I’m only checking for bit flips or simple file integrity issues.

Would it be possible to add a lightweight CLI interface to this project for simpler use cases like mine? I’d love to hear your thoughts or suggestions!

**Comment by hoxxep** [comment]
I have just [deployed `v1.2.0`](https://crates.io/crates/rapidhash) which should allow you to `cargo install rapidhash` and use `rapidhash` on the commandline (accepting both stdin or a file path). It doesn't have a `--help` flag yet though, I'll add that in a future release.

```sh
echo "example" | rapidhash 
8543579700415218186

rapidhash example.txt
8543579700415218186
```

**Comment by hoxxep** [comment]
Just tested on a 23G file on a local Macbook and rapidhash is IO bound, while sha512 is CPU bound. Happy hashing!
```
sha512      18.42s user 4.41s system 99% cpu 22.847 total
rapidhash   1.21s user 5.88s system 77% cpu 9.145 total
```

**Comment by hoxxep** [comment]
Yeah the CLI is a lazy implementation and reads the whole file because I simply copied the old CLI example, let me rework it to stream the file and add the `--help` flag. Brb!

**Comment by hoxxep** [comment]
Actually on second thought, the problem is that rapidhash doesn't have a streaming implementation. I'll have a think about whether I can come up with an alternative. You can read here for more info: https://github.com/Nicoshev/rapidhash/issues/3

**Comment by hoxxep** [comment]
@lucasbalena I've added a file-specific streaming implementation, currently on another branch. The pull request is: https://github.com/hoxxep/rapidhash/pull/10

Would you be willing to test the branch, to check the performance and memory usage matches what we are expecting?

You should be able to install the streaming branch via:
```sh
cargo install --git https://github.com/hoxxep/rapidhash --branch streaming rapidhash
```

Results I'm seeing on an M1 Max MacBook, with rapidhash using ~1MB ram, and b3sum using up to ~10MB. rapidhash is single-threaded in design and does significantly less CPU work, so I've included both single and multithreaded b3sum commands.
```sh
rapidhash                1.84s user  2.73s system   99% cpu   4.568 total 
b3sum                   16.76s user  2.31s system  844% cpu   2.259 total
b3sum --num-threads 1   14.77s user  3.34s system   99% cpu  18.124 total
```

**Comment by hoxxep** [comment]
Going to close this issue since the CLI and file streaming has been merged and released for a while now. The only thing I can think is perhaps checking the read length with the metadata length, happy to reopen if there's anything else. Cheers!
